File: frontend/src/types/sale.ts
export const DOCUMENT_CATEGORIES = [
  'groceries',
  'dining',
  'transport',
  'utilities',
  'health',
  'shopping',
  'entertainment',
  'travel',
  'housing',
  'other',
] as const
export type DocumentCategory = (typeof DOCUMENT_CATEGORIES)[number]

export const PAYMENT_METHODS = ['card', 'cash', 'other'] as const
export type PaymentMethod = (typeof PAYMENT_METHODS)[number]

// The closed set of transaction currencies selectable through the Sale
// form — see backend app/domain/demo_business.py. Currency is never the
// same thing as tax jurisdiction: this demo business is always taxed under
// Israeli VAT rules regardless of which of these a given sale is charged in.
export const TRANSACTION_CURRENCIES = ['ILS', 'USD', 'EUR'] as const
export type TransactionCurrency = (typeof TRANSACTION_CURRENCIES)[number]

// How Israeli VAT applies to a sale — see backend app/models/sale.py's
// TaxTreatment docstring. `standard` is the default; `zero_rate` and
// `exempt` both always produce vat_amount = 0.
export const TAX_TREATMENTS = ['standard', 'zero_rate', 'exempt'] as const
export type TaxTreatment = (typeof TAX_TREATMENTS)[number]

export const SALE_SOURCES = ['manual', 'csv', 'webhook', 'demo'] as const
export type SaleSource = (typeof SALE_SOURCES)[number]

export const SALE_STATUSES = ['succeeded', 'pending', 'failed', 'refunded', 'partially_refunded'] as const
export type SaleStatus = (typeof SALE_STATUSES)[number]

// `waiting_automatic` is distinct from `pending`: both mean "no document
// yet," but `pending` is a sale with no automatic document path at all
// (manual/CSV/no provider integration), while `waiting_automatic` is a
// sale whose payment provider (Grow/Cardcom) is expected to supply a
// document automatically and simply hasn't yet — see backend
// app/models/sale.py's DocumentStatus docstring.
export const DOCUMENT_STATUSES = ['pending', 'waiting_automatic', 'issued', 'failed', 'not_required'] as const
export type DocumentStatus = (typeof DOCUMENT_STATUSES)[number]

export interface Sale {
  id: string
  external_id: string | null
  source_provider: string | null
  source: SaleSource
  status: SaleStatus
  occurred_at: string
  customer_name: string
  customer_contact: string | null
  service_name: string
  description: string | null
  gross_amount: number | string
  vat_amount: number | string | null
  // Null only for a legacy sale a data migration couldn't safely classify
  // (see the backend migration's docstring) — every sale created going
  // forward always has a concrete value.
  tax_treatment: TaxTreatment | null
  tax_treatment_needs_review: boolean
  vat_rate: number | string | null
  processing_fee: number | string | null
  net_amount: number | string
  refunded_amount: number | string | null
  currency: string
  payment_method: string | null
  document_status: DocumentStatus
  document_number: string | null
  document_url: string | null
  // The provider's own document type label (e.g. Cardcom's
  // "TaxInvoiceAndReceipt"). Always null for Grow — its invoice webhook
  // never sends a type — and for any manually attached document.
  document_type: string | null
  raw_description: string | null
  created_at: string
  updated_at: string
}

export interface SaleInput {
  customer_name: string
  customer_contact?: string | null
  service_name: string
  description?: string | null
  gross_amount: number
  // vat_amount is deliberately absent: it is always backend-computed from
  // gross_amount + tax_treatment, never client-supplied — see SaleForm.
  tax_treatment?: TaxTreatment
  processing_fee?: number | null
  currency: TransactionCurrency | string
  payment_method?: PaymentMethod | null
  occurred_at: string
}

export interface SaleFilters {
  search?: string
  status?: SaleStatus | ''
  date_from?: string
  date_to?: string
}

File: backend/app/schemas/sale.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.domain.demo_business import DEMO_DEFAULT_TRANSACTION_CURRENCY
from app.models.sale import DocumentStatus, Sale, SaleSource, SaleStatus, TaxTreatment
from app.schemas.validators import (
    validate_finite_decimal,
    validate_payment_method,
    validate_required_text,
    validate_tax_treatment,
    validate_transaction_currency,
    validate_transaction_datetime_reasonable,
)


class SaleBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    customer_name: str = Field(min_length=1, max_length=255)
    customer_contact: str | None = Field(default=None, max_length=255)
    service_name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    gross_amount: Decimal = Field(ge=0, max_digits=12, decimal_places=2)
    # `vat_amount` is deliberately NOT a field here: it is always backend-
    # computed from gross_amount + tax_treatment (see app/services/tax/vat.py
    # and app/api/routes/sales.py) — never client-supplied, so it can never
    # conflict with the selected tax treatment. SaleRead below exposes the
    # computed result.
    tax_treatment: TaxTreatment = Field(default=TaxTreatment.STANDARD)
    processing_fee: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)
    currency: str = Field(default=DEMO_DEFAULT_TRANSACTION_CURRENCY, min_length=3, max_length=3)
    payment_method: str | None = Field(default=None, max_length=50)
    occurred_at: datetime

    @field_validator("customer_name", "service_name")
    @classmethod
    def _validate_required_text(cls, value: str) -> str:
        return validate_required_text(value)

    @field_validator("currency")
    @classmethod
    def _validate_currency(cls, value: str) -> str:
        return validate_transaction_currency(value)

    @field_validator("tax_treatment")
    @classmethod
    def _validate_tax_treatment(cls, value: TaxTreatment) -> TaxTreatment:
        validate_tax_treatment(value.value if isinstance(value, TaxTreatment) else value)
        return value

    @field_validator("gross_amount", "processing_fee")
    @classmethod
    def _validate_finite(cls, value: Decimal | None) -> Decimal | None:
        return validate_finite_decimal(value)

    @field_validator("payment_method")
    @classmethod
    def _validate_payment_method(cls, value: str | None) -> str | None:
        return validate_payment_method(value)

    @field_validator("occurred_at")
    @classmethod
    def _validate_occurred_at(cls, value: datetime) -> datetime:
        return validate_transaction_datetime_reasonable(value)  # type: ignore[return-value]


class SaleCreate(SaleBase):
    """Manual sale entry — the fallback path when no payment provider sent
    a webhook. Always recorded as a completed (`succeeded`) sale; `net_amount`
    and `vat_amount` are never accepted from the client, only computed
    server-side."""

    pass


class SaleUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    customer_name: str | None = Field(default=None, min_length=1, max_length=255)
    customer_contact: str | None = Field(default=None, max_length=255)
    service_name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=2000)
    gross_amount: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)
    tax_treatment: TaxTreatment | None = Field(default=None)
    processing_fee: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    payment_method: str | None = Field(default=None, max_length=50)
    occurred_at: datetime | None = None

    @field_validator("customer_name", "service_name")
    @classmethod
    def _validate_required_text(cls, value: str | None) -> str | None:
        return None if value is None else validate_required_text(value)

    @field_validator("currency")
    @classmethod
    def _validate_currency(cls, value: str | None) -> str | None:
        return None if value is None else validate_transaction_currency(value)

    @field_validator("tax_treatment")
    @classmethod
    def _validate_tax_treatment(cls, value: TaxTreatment | None) -> TaxTreatment | None:
        if value is None:
            return None
        validate_tax_treatment(value.value if isinstance(value, TaxTreatment) else value)
        return value

    @field_validator("gross_amount", "processing_fee")
    @classmethod
    def _validate_finite(cls, value: Decimal | None) -> Decimal | None:
        return validate_finite_decimal(value)

    @field_validator("payment_method")
    @classmethod
    def _validate_payment_method(cls, value: str | None) -> str | None:
        return validate_payment_method(value)

    @field_validator("occurred_at")
    @classmethod
    def _validate_occurred_at(cls, value: datetime | None) -> datetime | None:
        return None if value is None else validate_transaction_datetime_reasonable(value)  # type: ignore[return-value]


class SaleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    external_id: str | None = None
    source_provider: str | None = None
    source: SaleSource
    status: SaleStatus
    occurred_at: datetime
    customer_name: str
    customer_contact: str | None = None
    service_name: str
    description: str | None = None
    gross_amount: Decimal
    vat_amount: Decimal | None = None
    # Nullable because a legacy sale migrated before this field existed may
    # have an ambiguous history the migration couldn't safely resolve — see
    # `tax_treatment_needs_review` and the migration's docstring. Every sale
    # created going forward always has a concrete value.
    tax_treatment: TaxTreatment | None = None
    tax_treatment_needs_review: bool = False
    vat_rate: Decimal | None = None
    processing_fee: Decimal | None = None
    net_amount: Decimal
    refunded_amount: Decimal | None = None
    currency: str
    payment_method: str | None = None
    document_status: DocumentStatus
    document_number: str | None = None
    document_url: str | None = None
    raw_description: str | None = None
    created_at: datetime
    updated_at: datetime


class RefundRequest(BaseModel):
    """A manual refund record — no real payment provider sends a refund
    webhook in this demo, so refunds are recorded explicitly. Omitting
    `amount` records a full refund."""

    model_config = ConfigDict(extra="forbid")

    amount: Decimal | None = Field(default=None, gt=0, max_digits=12, decimal_places=2)

    @field_validator("amount")
    @classmethod
    def _validate_finite(cls, value: Decimal | None) -> Decimal | None:
        return validate_finite_decimal(value)


def sale_to_read(sale: Sale) -> SaleRead:
    return SaleRead.model_validate(sale)
